Language/Haskell/GhcMod/ImportedFrom.hs

```diff
@@ -0,0 +1,785 @@
+
+data ModuleExports = ModuleExports
+    { mName            :: StrModuleName            -- ^ e.g. "Data.List"
+    , mPackageName     :: String                   -- ^ e.g. "snap-0.14.0.6"
+    , mInfo            :: NiceImportDecl           -- ^ Our parse of the module import, with info like "hiding (map)".
+    , qualifiedExports :: [FullyQualifiedName]     -- ^ e.g. [ "base-4.8.2.0:GHC.Base.++"
+                                                   --        , "base-4.8.2.0:GHC.List.filter"
+                                                   --        , "base-4.8.2.0:GHC.List.zip"
+                                                   --        , ...
+                                                   --        ]
+    }
+    deriving Show
+
+refineAs :: MySymbol -> [ModuleExports] -> [ModuleExports]
+
+-- User qualified the symbol, so we can filter out anything that doesn't have a matching 'modImportedAs'.
+refineAs (MySymbolUserQualified userQualSym) exports = filter f exports
+  where
+    f export = case modas of
+                Nothing     -> False
+                Just modas' -> modas' == userQualAs
+       where modas = modImportedAs $ mInfo export :: Maybe String
+
+             -- e.g. "DL"
+             userQualAs = fromMaybe (error $ "Expected a qualified name like 'DL.length' but got: " ++ userQualSym)
+                                    (moduleOfQualifiedName userQualSym)
+
+-- User didn't qualify the symbol, so we have the full system qualified thing, so do nothing here.
+refineAs (MySymbolSysQualified _) exports = exports
+
+refineRemoveHiding :: [ModuleExports] -> [ModuleExports]
+refineRemoveHiding exports = map (\e -> e { qualifiedExports = f e }) exports
+  where
+    f export = filter (`notElem` hiding') thisExports
+       where hiding = modHiding $ mInfo export :: [String] -- Things that this module hides.
+             hiding' = map (qualifyName thisExports) hiding  :: [String]    -- Qualified version of hiding.
+             thisExports = qualifiedExports export         -- Things that this module exports.
+
+    qualifyName :: [QualifiedName] -> String -> QualifiedName
+    qualifyName qualifiedNames name
+        -- = case filter (postfixMatch name) qualifiedNames of
+        = case nub (filter (name `f`) qualifiedNames) of
+            [match]     -> match
+            m           -> error $ "Could not qualify " ++ name ++ " from these exports: " ++ show qualifiedNames ++ "\n    matches: " ++ show m
+
+        -- Time for some stringly typed rubbish. The previous test used
+        -- postfixMatch but this failed on an import that had "hiding (lines, unlines)" since
+        -- both lines and unlines matched. Prepending a dot doesn't work due to things like ".=" from
+        -- Control.Lens. So we manually check that the suffix matches, that the next symbol is a dot,
+        -- and then an alpha character, which hopefully is the end of a module name. Such a mess.
+        where f n qn = if length qn - length n - 2 >= 0
```

Comment:
If you're trying to parse a `hiding` list here maybe you'd be better served by using ReadP rather than this abomination? If you give me some examples what the input looks like I could do it for you.